homeassistant/components/sensor/gtfs.py

```diff
@@ -0,0 +1,249 @@
+        "Drop Off Type": item[13], "Pickup Type": item[14],
+        "Shape Dist Traveled": item[15], "Headsign": item[16],
+        "Sequence": item[17]
+    }
+
+    return {
+        "trip_id": item[0],
+        "trip": sched.trips_by_id(item[0])[0],
+        "route": route,
+        "agency": sched.agencies_by_id(route.agency_id)[0],
+        "origin_station": origin_station,
+        "departure_time": departure_time,
+        "destination_station": destination_station,
+        "arrival_time": arrival_time,
+        "seconds_until_departure": seconds_until,
+        "minutes_until_departure": minutes_until,
+        "origin_stop_time": origin_stop_time_dict,
+        "destination_stop_time": destination_stop_time_dict
+    }
+
+
+def setup_platform(hass, config, add_devices, discovery_info=None):
+    """Get the GTFS sensor."""
+    if config.get("origin") is None:
+        _LOGGER.error("Origin must be set in the GTFS configuration!")
+        return False
+
+    if config.get("destination") is None:
+        _LOGGER.error("Destination must be set in the GTFS configuration!")
+        return False
+
+    if config.get("data") is None:
+        _LOGGER.error("Data must be set in the GTFS configuration!")
+        return False
+
+    dev = []
+    dev.append(GTFSDepartureSensor(config["data"], hass.config.path("gtfs"),
+                                   config["origin"], config["destination"]))
+    add_devices(dev)
+
+# pylint: disable=too-many-instance-attributes,too-few-public-methods
+
+
+class GTFSDepartureSensor(Entity):
+    """Implementation of an GTFS departures sensor."""
+
+    def __init__(self, data_source, gtfs_folder, origin, destination):
+        """Initialize the sensor."""
+        self._data_source = data_source
+        self._gtfs_folder = gtfs_folder
+        self.origin = origin
+        self.destination = destination
+        self._name = "GTFS Sensor"
+        self._unit_of_measurement = "min"
+        self._state = 0
+        self._attributes = {}
+        self.update()
+
+    @property
+    def name(self):
+        """Return the name of the sensor."""
+        return self._name
+
+    @property
+    def state(self):
+        """Return the state of the sensor."""
+        return self._state
+
+    @property
+    def unit_of_measurement(self):
+        """Return the unit of measurement of this entity, if any."""
+        return self._unit_of_measurement
+
+    @property
+    def device_state_attributes(self):
+        """Return the state attributes."""
+        return self._attributes
+
+    @property
+    def icon(self):
+        """Icon to use in the frontend, if any."""
+        return ICON
+
+    def update(self):
+        """Get the latest data from GTFS and update the states."""
+        import pygtfs
+
+        split_file_name = os.path.splitext(self._data_source)
+
+        sqlite_file = "{}.sqlite".format(split_file_name[0])
+        gtfs = pygtfs.Schedule(os.path.join(self._gtfs_folder, sqlite_file))
```

Comment:
The data source is entirely read only and updates somewhat rarely (at the very most I would say once a day) and an update requires a restart so I'm not super worried about any of the above.